src/gpiocmds.c

```diff
@@ -33,11 +42,43 @@ static uint_fast8_t
 digital_out_event(struct timer *timer)
 {
     struct digital_out_s *d = container_of(timer, struct digital_out_s, timer);
-    gpio_out_write(d->pin, d->value);
-    if (d->value == d->default_value || !d->max_duration)
-        return SF_DONE;
-    d->timer.waketime += d->max_duration;
-    d->timer.func = digital_end_event;
+    struct mq_event *c = mq_event_pop(&d->queue);
+
+    struct do_event* current = container_of(c, struct do_event, event);
+    gpio_out_write(d->pin, current->value);
+    d->current_value = current->value;
+
+    //may be NULL if no further elements are queued
+    struct mq_event* next = mq_event_peek(&d->queue);
+
+    if(next) {
+        //next event scheduled
+        uint32_t next_wt = container_of(c, struct do_event, event)->waketime;
+        if (current->value != d->default_value || !d->max_duration) {
+            if(next_wt > d->timer.waketime + d->max_duration) {
+                shutdown("Scheduled digital_out event will miss max_duration");
+            }
+        }
+        d->timer.waketime = next_wt;
+    } else {
+        // no next event scheduled
+        if (current->value == d->default_value || !d->max_duration) {
+            // We either have set the default value
+            // or there is no maximum duration
+            irqstatus_t irq = irq_save();
+            mq_free_event(current);
+            irq_restore(irq);
```

Comment:
Minor, but "software timers" are always called with irqs disabled, so the irq_save()/irq_restore() isn't needed.